File: scripts/build-player-sheet.py
#!/usr/bin/env python3
"""Build player-sheet.png from per-animation sprite strips."""

from __future__ import annotations

from pathlib import Path

try:
    from PIL import Image
except ImportError:
    raise SystemExit('Pillow required: pip install pillow')

ROOT = Path(__file__).resolve().parents[1]
SPRITE_DIR = ROOT / 'public/assets/sprite'
OUT_PATH = ROOT / 'public/assets/images/player-sheet.png'

ANIMATIONS: list[tuple[str, int, str | None, list[int] | None]] = [
    ('idle', 8, 'player-idle.png', None),
    ('run', 6, 'player-run.png', None),
    ('jump', 1, 'player-jump.png', [3]),
    ('fall', 1, 'player-fall.png', [2]),
    ('attack', 2, 'player-attack.png', [2, 3]),
]


def load_strip(
    path: Path,
    expected_frames: int,
    fallback: Image.Image,
    frame_indices: list[int] | None = None,
) -> list[Image.Image]:
    if not path.exists():
        return [fallback.copy() for _ in range(expected_frames)]

    sheet = Image.open(path).convert('RGBA')
    frame_count = max(1, sheet.width // sheet.height)
    frame_w = sheet.width // frame_count
    frame_h = sheet.height

    frames = [
        sheet.crop((index * frame_w, 0, (index + 1) * frame_w, frame_h))
        for index in range(frame_count)
    ]

    if frame_indices is not None:
        selected = [frames[min(index, len(frames) - 1)] for index in frame_indices]
        if len(selected) >= expected_frames:
            return selected[:expected_frames]
        while len(selected) < expected_frames:
            selected.append(selected[-1].copy())
        return selected

    if len(frames) >= expected_frames:
        return frames[:expected_frames]

    while len(frames) < expected_frames:
        frames.append(frames[-1].copy())

    return frames


def main() -> None:
    idle_path = SPRITE_DIR / 'player-idle.png'
    if not idle_path.exists():
        raise SystemExit(f'Missing required idle strip: {idle_path}')

    idle_sheet = Image.open(idle_path).convert('RGBA')
    idle_frame_w = idle_sheet.width // 8
    idle_frame_h = idle_sheet.height
    placeholder = idle_sheet.crop((0, 0, idle_frame_w, idle_frame_h))

    all_frames: list[Image.Image] = []
    for name, expected_frames, file_name, frame_indices in ANIMATIONS:
        source_path = SPRITE_DIR / file_name if file_name else idle_path
        strip_frames = load_strip(source_path, expected_frames, placeholder, frame_indices)
        all_frames.extend(strip_frames)
        status = 'ok' if source_path.exists() else 'placeholder'
        print(f'{name}: {len(strip_frames)} frames ({status})')

    frame_w = all_frames[0].width
    frame_h = all_frames[0].height
    out = Image.new('RGBA', (frame_w * len(all_frames), frame_h))

    for index, frame in enumerate(all_frames):
        if frame.size != (frame_w, frame_h):
            frame = frame.resize((frame_w, frame_h), Image.NEAREST)
        out.paste(frame, (index * frame_w, 0))

    OUT_PATH.parent.mkdir(parents=True, exist_ok=True)
    out.save(OUT_PATH)
    print(
        f'Wrote {OUT_PATH} ({out.size[0]}x{out.size[1]}, '
        f'{len(all_frames)} frames, {frame_w}x{frame_h} each)'
    )


if __name__ == '__main__':
    main()

File: scripts/build-player-sheet.mjs
#!/usr/bin/env node
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';
import sharp from 'sharp';

const __dirname = path.dirname(fileURLToPath(import.meta.url));
const ROOT = path.resolve(__dirname, '..');
const SPRITE_DIR = path.join(ROOT, 'public/assets/sprite');
const OUT_PATH = path.join(ROOT, 'public/assets/images/player-sheet.png');

const ANIMATIONS = [
  { name: 'idle', count: 8, file: 'player-idle.png', indices: null },
  { name: 'run', count: 6, file: 'player-run.png', indices: null },
  { name: 'jump', count: 1, file: 'player-jump.png', indices: [3] },
  { name: 'fall', count: 1, file: 'player-fall.png', indices: [2] },
  { name: 'attack', count: 2, file: 'player-attack.png', indices: [2, 3] },
];

async function loadFrames(file, count, indices) {
  const filePath = path.join(SPRITE_DIR, file);
  const meta = await sharp(filePath).metadata();
  const frameCount = Math.max(1, Math.floor(meta.width / meta.height));
  const frameW = Math.floor(meta.width / frameCount);
  const frameH = meta.height;
  const pick = indices ?? Array.from({ length: count }, (_, index) => index);
  const frames = [];

  for (const index of pick.slice(0, count)) {
    const safeIndex = Math.min(index, frameCount - 1);
    frames.push(
      await sharp(filePath)
        .extract({ left: safeIndex * frameW, top: 0, width: frameW, height: frameH })
        .png()
        .toBuffer(),
    );
  }

  while (frames.length < count) {
    frames.push(frames[frames.length - 1]);
  }

  return frames;
}

async function main() {
  const allFrames = [];

  for (const animation of ANIMATIONS) {
    const frames = await loadFrames(animation.file, animation.count, animation.indices);
    allFrames.push(...frames);
    console.log(`${animation.name}: ${frames.length} frames`);
  }

  const firstMeta = await sharp(allFrames[0]).metadata();
  const frameW = firstMeta.width ?? 172;
  const frameH = firstMeta.height ?? 172;

  const composites = allFrames.map((buffer, index) => ({
    input: buffer,
    left: index * frameW,
    top: 0,
  }));

  fs.mkdirSync(path.dirname(OUT_PATH), { recursive: true });
  await sharp({
    create: {
      width: frameW * allFrames.length,
      height: frameH,
      channels: 4,
      background: { r: 0, g: 0, b: 0, alpha: 0 },
    },
  })
    .composite(composites)
    .png()
    .toFile(OUT_PATH);

  console.log(
    `Wrote ${OUT_PATH} (${frameW * allFrames.length}x${frameH}, ${allFrames.length} frames)`,
  );
}

main().catch((error) => {
  console.error(error);
  process.exit(1);
});
